File: apps/api/domain/use_cases/generate_inspiration.py
"""Use case: the AI inspiration pipeline.

keywords → Ad Library → (Higgsfield images ∥ Claude copies) → persist job state.

This use case is pure: it orchestrates ports only and knows nothing about
Redis, HTTP, Higgsfield or Anthropic concretely.
"""
from __future__ import annotations

import asyncio
from dataclasses import dataclass, field

from domain.models.inspiration import GeneratedImage, InspirationJob, JobStatus
from domain.ports.outbound.ad_library import IAdLibrary
from domain.ports.outbound.copy_generator import ICopyGenerator
from domain.ports.outbound.image_generator import IImageGenerator
from domain.ports.outbound.job_store import IJobStore


@dataclass
class GenerateInspirationCommand:
    job_id: str
    client_id: str
    client_name: str
    product: str
    description: str
    objective: str
    country: str
    webhook_base_url: str
    platforms: list[str] = field(default_factory=lambda: ["facebook", "instagram"])
    image_count: int = 5
    copy_count: int = 10


class GenerateInspirationUseCase:
    def __init__(
        self,
        ad_library: IAdLibrary,
        image_generator: IImageGenerator,
        copy_generator: ICopyGenerator,
        job_store: IJobStore,
    ):
        self.ad_library = ad_library
        self.image_generator = image_generator
        self.copy_generator = copy_generator
        self.job_store = job_store

    async def execute(self, cmd: GenerateInspirationCommand) -> InspirationJob:
        job = await self.job_store.get(cmd.job_id) or InspirationJob(
            id=cmd.job_id, client_id=cmd.client_id, country=cmd.country
        )

        try:
            # 1. keywords + Ad Library research
            job.status = JobStatus.SEARCHING
            job.platforms = cmd.platforms
            await self.job_store.save(job)

            keywords = await self.copy_generator.extract_keywords(cmd.description)
            job.keywords = keywords
            reference_ads = await self.ad_library.search_top_ads(
                keywords, cmd.country, cmd.platforms
            )
            job.reference_ads = reference_ads

            # 2. images + copies in parallel
            job.status = JobStatus.GENERATING
            await self.job_store.save(job)

            webhook_url = f"{cmd.webhook_base_url}/webhooks/higgsfield/{job.id}"
            request_ids, copies = await asyncio.gather(
                self.image_generator.generate(
                    reference_ads, cmd.product, cmd.country, webhook_url, cmd.image_count
                ),
                self.copy_generator.generate_copies(
                    reference_ads,
                    cmd.client_name,
                    cmd.product,
                    cmd.objective,
                    cmd.country,
                    cmd.copy_count,
                ),
            )
            job.higgsfield_request_ids = request_ids
            job.assets.copies = copies
            # Stub image generators return ready URLs directly; the Higgsfield
            # webhook fills these in for real generation.
            job.assets.images = [
                GeneratedImage(url=rid, request_id=rid)
                for rid in request_ids
                if rid.startswith("http")
            ]

            job.status = JobStatus.READY
            await self.job_store.save(job)
        except Exception as exc:  # noqa: BLE001 — record failure on the job
            job.status = JobStatus.FAILED
            job.error = str(exc)
            await self.job_store.save(job)
            raise

        return job

File: apps/api-go/internal/domain/usecase/generate_inspiration.go
package usecase

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/model"
	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/port"
)

// GenerateInspirationCommand is the input for the AI pipeline.
type GenerateInspirationCommand struct {
	JobID          string
	ClientID       string
	ClientName     string
	Product        string
	Description    string
	Objective      string
	Country        string
	WebhookBaseURL string
	Platforms      []string
	ImageCount     int
	CopyCount      int
}

// GenerateInspirationUseCase runs: keywords → Ad Library → (images ∥ copies).
type GenerateInspirationUseCase struct {
	adLibrary port.AdLibrary
	imageGen  port.ImageGenerator
	copyGen   port.CopyGenerator
	jobStore  port.JobStore
}

func NewGenerateInspirationUseCase(
	adLibrary port.AdLibrary,
	imageGen port.ImageGenerator,
	copyGen port.CopyGenerator,
	jobStore port.JobStore,
) *GenerateInspirationUseCase {
	return &GenerateInspirationUseCase{adLibrary: adLibrary, imageGen: imageGen, copyGen: copyGen, jobStore: jobStore}
}

// Execute drives the pipeline, persisting the job after each phase so the
// frontend's polling sees progress (searching → generating → ready).
func (uc *GenerateInspirationUseCase) Execute(ctx context.Context, cmd GenerateInspirationCommand) (model.InspirationJob, error) {
	job, err := uc.jobStore.Get(ctx, cmd.JobID)
	if err != nil {
		// No existing job: start a fresh one.
		job = model.InspirationJob{
			ID:        cmd.JobID,
			ClientID:  cmd.ClientID,
			Country:   cmd.Country,
			Platforms: cmd.Platforms,
			CreatedAt: time.Now().UTC(),
		}
	}

	// fail records the error on the job, persists it, and returns it. A small
	// closure keeps the error-handling DRY across the phases below.
	fail := func(err error) (model.InspirationJob, error) {
		job.Status = model.JobFailed
		job.Error = err.Error()
		_ = uc.jobStore.Save(ctx, job)
		return job, err
	}

	// ── Phase 1: keywords + Ad Library research ──────────────────────────
	job.Status = model.JobSearching
	if err := uc.jobStore.Save(ctx, job); err != nil {
		return fail(fmt.Errorf("save job: %w", err))
	}
	keywords, err := uc.copyGen.ExtractKeywords(ctx, cmd.Description)
	if err != nil {
		return fail(fmt.Errorf("extract keywords: %w", err))
	}
	job.Keywords = keywords
	refs, err := uc.adLibrary.SearchTopAds(ctx, keywords, cmd.Country, cmd.Platforms, 5)
	if err != nil {
		return fail(fmt.Errorf("search ad library: %w", err))
	}
	job.ReferenceAds = refs

	// ── Phase 2: images ∥ copies (concurrent) ────────────────────────────
	//
	// Go concurrency: we launch two goroutines and wait for both with a
	// WaitGroup. Each goroutine writes to its OWN result variables, so there is
	// no shared mutable state and therefore no data race — no mutex needed.
	job.Status = model.JobGenerating
	if err := uc.jobStore.Save(ctx, job); err != nil {
		return fail(fmt.Errorf("save job: %w", err))
	}
	webhookURL := fmt.Sprintf("%s/webhooks/higgsfield/%s", strings.TrimRight(cmd.WebhookBaseURL, "/"), job.ID)

	var (
		wg         sync.WaitGroup
		requestIDs []string
		copies     []model.Copy
		imgErr     error
		copyErr    error
	)
	wg.Add(2)
	go func() {
		defer wg.Done()
		requestIDs, imgErr = uc.imageGen.Generate(ctx, refs, cmd.Product, cmd.Country, webhookURL, cmd.ImageCount)
	}()
	go func() {
		defer wg.Done()
		copies, copyErr = uc.copyGen.GenerateCopies(ctx, refs, cmd.ClientName, cmd.Product, cmd.Objective, cmd.Country, cmd.CopyCount)
	}()
	wg.Wait()

	if imgErr != nil {
		return fail(fmt.Errorf("generate images: %w", imgErr))
	}
	if copyErr != nil {
		return fail(fmt.Errorf("generate copies: %w", copyErr))
	}

	job.HiggsfieldRequestIDs = requestIDs
	job.Assets.Copies = copies
	// Stub generators return ready https URLs directly; the real Higgsfield flow
	// returns opaque request IDs and the webhook fills the URLs in later.
	for _, rid := range requestIDs {
		if strings.HasPrefix(rid, "http") {
			job.Assets.Images = append(job.Assets.Images, model.GeneratedImage{URL: rid, RequestID: rid})
		}
	}

	job.Status = model.JobReady
	if err := uc.jobStore.Save(ctx, job); err != nil {
		return fail(fmt.Errorf("save job: %w", err))
	}
	return job, nil
}
